sql/core/src/main/scala/org/apache/spark/sql/execution/python/TransformWithStateInPandasPythonRunner.scala

```diff
@@ -0,0 +1,175 @@
+
+  private val arrowWriter: arrow.ArrowWriter = ArrowWriter.create(root)
+
+  // Use lazy val to initialize the fields before these are accessed in [[PythonArrowInput]]'s
+  // constructor.
+  override protected lazy val schema: StructType = _schema
+  override protected lazy val timeZoneId: String = _timeZoneId
+  override protected val errorOnDuplicatedFieldNames: Boolean = true
+  override protected val largeVarTypes: Boolean = sqlConf.arrowUseLargeVarTypes
+
+  override protected def handleMetadataBeforeExec(stream: DataOutputStream): Unit = {
+    super.handleMetadataBeforeExec(stream)
+    // Also write the port number for state server
+    stream.writeInt(serverId)
+  }
+
+  override def compute(
+      inputIterator: Iterator[InType],
+      partitionIndex: Int,
+      context: TaskContext): Iterator[OutType] = {
+    var serverChannel: UnixServerSocketChannel = null
+    var failed = false
+    try {
+      val socketFile = Path.of(socketPath)
+      Files.deleteIfExists(socketFile)
+      val serverAddress = new UnixSocketAddress(socketPath)
+      serverChannel = UnixServerSocketChannel.open()
+      serverChannel.socket().bind(serverAddress)
+    } catch {
+      case e: Exception =>
+        failed = true
+        throw e
+    } finally {
+      if (failed) {
+        closeServerSocketChannelSilently(serverChannel)
+      }
+    }
+
+    val executor = ThreadUtils.newDaemonSingleThreadExecutor("stateConnectionListenerThread")
+    val executionContext = ExecutionContext.fromExecutor(executor)
+
+    executionContext.execute(
+      new TransformWithStateInPandasStateServer(serverChannel, processorHandle,
+        groupingKeySchema))
+
+    context.addTaskCompletionListener[Unit] { _ =>
+      logWarning(s"completion listener called")
+      executor.awaitTermination(10, TimeUnit.SECONDS)
+      executor.shutdownNow()
+      val socketFile = Path.of(socketPath)
+      Files.deleteIfExists(socketFile)
+    }
+
+    super.compute(inputIterator, partitionIndex, context)
+  }
+
+  private def closeServerSocketChannelSilently(serverChannel: UnixServerSocketChannel): Unit = {
+    try {
+      logWarning(s"closing the state server socket")
+      serverChannel.close()
+    } catch {
+      case e: Exception =>
+        logError(s"failed to close state server socket", e)
+    }
+  }
+
+  override protected def writeUDF(dataOut: DataOutputStream): Unit = {
+    PythonUDFRunner.writeUDFs(dataOut, funcs, argOffsets, None)
+  }
+
+  override protected def writeNextInputToArrowStream(
+      root: VectorSchemaRoot,
+      writer: ArrowStreamWriter,
+      dataOut: DataOutputStream,
+      inputIterator: Iterator[InType]): Boolean = {
+
+    if (inputIterator.hasNext) {
+      val startData = dataOut.size()
+      val next = inputIterator.next()
+      val nextBatch = next._2
+
+      while (nextBatch.hasNext) {
+        arrowWriter.write(nextBatch.next())
+      }
```

Comment:
This is sending the entire data for a key to Python client, which would result in data for entire grouping key to be buffered in memory. We should chunk it (similar to ApplyInPandasWithStatePythonRunner).

Reply:
As we discussed, this work will be captured in a followup PR.